File: src/services/dashboardStories.ts
import type {
  DiscoveredDeal,
  Retailer,
  StoreLeaflet,
} from '../types'

export interface DashboardStoryFrame {
  catalogue?: StoreLeaflet
  deal?: DiscoveredDeal
  id: string
  imageUrl: string
  imageUrls: string[]
  kind: 'catalogue' | 'deal'
  pageNumber?: number
  sourceUrl: string
  subtitle?: string
  title: string
}

export interface DashboardStory {
  frames: DashboardStoryFrame[]
  id: string
  logoUrl?: string
  retailerName: string
}

const MAX_STORIES = 16
const MAX_FRAMES_PER_STORY = 40

interface MutableStory {
  catalogues: DashboardStoryFrame[]
  deals: DashboardStoryFrame[]
  id: string
  retailerName: string
}

export function buildDashboardStories(
  catalogues: StoreLeaflet[],
  deals: DiscoveredDeal[],
  retailers: Array<Pick<Retailer, 'id' | 'logoUrl' | 'name'>>,
): DashboardStory[] {
  const groups = new Map<string, MutableStory>()
  const order: string[] = []

  const storyFor = (id: string, name: string) => {
    const existing = groups.get(id)
    if (existing) return existing
    if (groups.size >= MAX_STORIES) return undefined
    const created: MutableStory = {
      catalogues: [],
      deals: [],
      id,
      retailerName: name,
    }
    groups.set(id, created)
    order.push(id)
    return created
  }

  for (const catalogue of catalogues) {
    const group = storyFor(catalogue.retailerId, catalogue.retailerName)
    if (!group || group.catalogues.length >= MAX_FRAMES_PER_STORY) continue
    const pages = [...(catalogue.pages ?? [])]
      .filter((page) => page.imageUrl.trim().length > 0)
      .sort((left, right) => left.pageNumber - right.pageNumber)

    if (pages.length > 0) {
      for (const page of pages.slice(
        0,
        MAX_FRAMES_PER_STORY - group.catalogues.length,
      )) {
        group.catalogues.push({
          catalogue,
          id: `${catalogue.id}:page:${page.pageNumber}`,
          imageUrl: page.imageUrl,
          imageUrls: unique([page.imageUrl, ...(page.fallbacks ?? [])]),
          kind: 'catalogue',
          pageNumber: page.pageNumber,
          sourceUrl: catalogue.url,
          subtitle: `Page ${page.pageNumber} of ${pages.length}`,
          title: catalogue.name,
        })
      }
    } else if (catalogue.imageUrl) {
      group.catalogues.push({
        catalogue,
        id: `${catalogue.id}:cover`,
        imageUrl: catalogue.imageUrl,
        imageUrls: [catalogue.imageUrl],
        kind: 'catalogue',
        pageNumber: 1,
        sourceUrl: catalogue.url,
        subtitle: 'Catalogue cover',
        title: catalogue.name,
      })
    }
  }

  for (const deal of deals) {
    const imageUrl = deal.imageUrl ?? deal.images?.find(Boolean)
    if (!imageUrl) continue
    const group = storyFor(deal.retailerId, deal.retailerName)
    if (
      !group ||
      group.catalogues.length + group.deals.length >= MAX_FRAMES_PER_STORY
    ) {
      continue
    }
    group.deals.push({
      deal,
      id: `deal:${deal.id}`,
      imageUrl,
      imageUrls: unique([imageUrl, ...(deal.images ?? [])]),
      kind: 'deal',
      sourceUrl: deal.productUrl,
      subtitle: deal.priceText,
      title: deal.title,
    })
  }

  const retailerLookup = new Map(retailers.map((retailer) => [retailer.id, retailer]))
  return order
    .map((id) => {
      const group = groups.get(id)!
      const retailer = retailerLookup.get(id)
      return {
        frames: [...group.catalogues, ...group.deals].slice(0, MAX_FRAMES_PER_STORY),
        id,
        logoUrl: retailer?.logoUrl,
        retailerName: retailer?.name ?? group.retailerName,
      }
    })
    .filter((story) => story.frames.length > 0)
    .slice(0, MAX_STORIES)
}

function unique(values: string[]): string[] {
  return [...new Set(values.map((value) => value.trim()).filter(Boolean))]
}

File: mobile/lib/dashboard_stories.dart
import 'api_models.dart';

enum DashboardStoryFrameKind { catalogue, deal }

class DashboardStoryFrame {
  const DashboardStoryFrame({
    required this.id,
    required this.kind,
    required this.imageUrl,
    required this.imageUrls,
    required this.title,
    required this.sourceUrl,
    this.subtitle,
    this.pageNumber,
    this.catalogue,
    this.deal,
  });

  final String id;
  final DashboardStoryFrameKind kind;
  final String imageUrl;
  final List<String> imageUrls;
  final String title;
  final String sourceUrl;
  final String? subtitle;
  final int? pageNumber;
  final Catalogue? catalogue;
  final Deal? deal;
}

class DashboardStory {
  const DashboardStory({
    required this.id,
    required this.retailerName,
    required this.frames,
    this.logoUrl,
  });

  final String id;
  final String retailerName;
  final String? logoUrl;
  final List<DashboardStoryFrame> frames;
}

class _MutableStory {
  _MutableStory({required this.id, required this.retailerName});

  final String id;
  final String retailerName;
  final catalogues = <DashboardStoryFrame>[];
  final deals = <DashboardStoryFrame>[];
}

List<DashboardStory> buildDashboardStories({
  required List<Catalogue> catalogues,
  required List<Deal> deals,
  required List<Retailer> retailers,
  int maxStories = 16,
  int maxFramesPerStory = 40,
}) {
  final retailerById = {
    for (final retailer in retailers) retailer.id: retailer
  };
  final retailerByName = {
    for (final retailer in retailers)
      retailer.name.trim().toLowerCase(): retailer
  };
  final groups = <String, _MutableStory>{};
  final order = <String>[];

  _MutableStory? storyFor(String id, String name) {
    final existing = groups[id];
    if (existing != null) return existing;
    if (groups.length >= maxStories) return null;
    return groups.putIfAbsent(id, () {
      order.add(id);
      return _MutableStory(id: id, retailerName: name);
    });
  }

  for (final catalogue in catalogues) {
    final retailerName = catalogue.retailerName?.trim().isNotEmpty == true
        ? catalogue.retailerName!.trim()
        : 'Store';
    final retailer = retailerByName[retailerName.toLowerCase()];
    final id = retailer?.id ?? 'name:${_slug(retailerName)}';
    final story = storyFor(id, retailerName);
    if (story == null || story.catalogues.length >= maxFramesPerStory) {
      continue;
    }
    final pages = [...catalogue.pages]
      ..removeWhere((page) => page.imageUrl.trim().isEmpty)
      ..sort((left, right) => left.pageNumber.compareTo(right.pageNumber));

    if (pages.isNotEmpty) {
      for (final page
          in pages.take(maxFramesPerStory - story.catalogues.length)) {
        story.catalogues.add(DashboardStoryFrame(
          id: '${catalogue.url}:page:${page.pageNumber}',
          kind: DashboardStoryFrameKind.catalogue,
          imageUrl: page.imageUrl,
          imageUrls: page.imageUrls,
          title: catalogue.name,
          sourceUrl: catalogue.sourceUrl ?? catalogue.url,
          subtitle: 'Page ${page.pageNumber} of ${pages.length}',
          pageNumber: page.pageNumber,
          catalogue: catalogue,
        ));
      }
    } else if (catalogue.coverImageUrl != null) {
      story.catalogues.add(DashboardStoryFrame(
        id: '${catalogue.url}:cover',
        kind: DashboardStoryFrameKind.catalogue,
        imageUrl: catalogue.coverImageUrl!,
        imageUrls: [catalogue.coverImageUrl!],
        title: catalogue.name,
        sourceUrl: catalogue.sourceUrl ?? catalogue.url,
        subtitle: 'Catalogue cover',
        pageNumber: 1,
        catalogue: catalogue,
      ));
    }
  }

  for (final deal in deals) {
    if (!deal.hasImage) continue;
    final id = deal.retailerId.trim().isEmpty
        ? 'name:${_slug(deal.retailerName)}'
        : deal.retailerId;
    final story = storyFor(id, deal.retailerName);
    if (story == null ||
        story.catalogues.length + story.deals.length >= maxFramesPerStory) {
      continue;
    }
    story.deals.add(DashboardStoryFrame(
      id: 'deal:${deal.id}:${deal.productUrl ?? deal.sourceUrl}',
      kind: DashboardStoryFrameKind.deal,
      imageUrl: deal.gallery.first,
      imageUrls: deal.gallery,
      title: deal.title,
      sourceUrl: deal.productUrl ?? deal.sourceUrl,
      subtitle: deal.priceText,
      deal: deal,
    ));
  }

  return order
      .map((id) {
        final group = groups[id]!;
        final retailer = retailerById[id];
        return DashboardStory(
          id: id,
          retailerName: retailer?.name ?? group.retailerName,
          logoUrl: retailer?.logoUrl,
          frames: [...group.catalogues, ...group.deals]
              .take(maxFramesPerStory)
              .toList(growable: false),
        );
      })
      .where((story) => story.frames.isNotEmpty)
      .take(maxStories)
      .toList(growable: false);
}

String _slug(String value) => value
    .trim()
    .toLowerCase()
    .replaceAll(RegExp(r'[^a-z0-9]+'), '-')
    .replaceAll(RegExp(r'^-+|-+$'), '');
